connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```diff
@@ -1235,6 +1244,246 @@ void sourceConnectorOffsets(String connName, ConnectorOffsetBackingStore offsetS
+                            .collect(Collectors.toSet());
+
+                    if (!partitionsToReset.isEmpty()) {
+                        log.debug("Deleting the consumer group offsets for the following topic partitions using an admin client for sink connector {}: {}.",
+                                connName, partitionsToReset);
+                        DeleteConsumerGroupOffsetsOptions deleteConsumerGroupOffsetsOptions = new DeleteConsumerGroupOffsetsOptions().timeoutMs(
+                                (int) ConnectResource.DEFAULT_REST_REQUEST_TIMEOUT_MS);
+                        DeleteConsumerGroupOffsetsResult deleteConsumerGroupOffsetsResult = admin.deleteConsumerGroupOffsets(groupId, partitionsToReset,
+                                deleteConsumerGroupOffsetsOptions);
+
+                        adminFuture = KafkaFuture.allOf(adminFuture, deleteConsumerGroupOffsetsResult.all());
+                    }
+
+                    adminFuture.whenComplete((ignored, error) -> {
+                        if (error != null) {
+                            // When a consumer group is non-empty, only group members can commit offsets. An attempt to alter offsets via the admin client
+                            // will result in an UnknownMemberIdException if the consumer group is non-empty (i.e. if the sink tasks haven't stopped
+                            // completely or if the connector is resumed while the alter offsets request is being processed). Similarly, an attempt to
+                            // delete consumer group offsets for a non-empty consumer group will result in a GroupSubscribedToTopicException
+                            if (error instanceof UnknownMemberIdException || error instanceof GroupSubscribedToTopicException) {
+                                cb.onCompletion(new ConnectException("Failed to alter consumer group offsets for connector " + connName + " either because its tasks " +
+                                                "haven't stopped completely yet or the connector was resumed before the request to alter its offsets could be successfully " +
+                                                "completed. If the connector is in a stopped state, this operation can be safely retried. If it doesn't eventually succeed, the " +
+                                                "Connect cluster may need to be restarted to get rid of the zombie sink tasks."),
+                                        null);
+                            } else {
+                                cb.onCompletion(new ConnectException("Failed to alter consumer group offsets for connector " + connName, error), null);
+                            }
+                        } else {
+                            completeAlterOffsetsCallback(alterOffsetsResult, cb);
+                        }
+                        Utils.closeQuietly(admin, "Offset alter admin for sink connector " + connName);
```

Comment:
This will leak the client if any of the preceding callback logic fails.

Can we perform this separately with another chained `whenComplete` invocation?